### ISSUE TITLE
const-propagated `arithmetic_overflow` in unreachable code

### ISSUE THREAD

**Issue by cuviper** [rust-lang/rust#109731]
### Code

I tried this code: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f6804071f4851d138db85474dbd096c8)

```rust
fn main() {
    let x = 2u32;
    let y = 3u32;
    // ...
    if y <= x {
        dbg!(x - y);
    }
}
```

I expected to see this happen: compiles, nothing printed

Instead, this happened:

```
error: this arithmetic operation will overflow
 --> src/main.rs:6:14
  |
6 |         dbg!(x - y);
  |              ^^^^^ attempt to compute `2_u32 - 3_u32`, which would overflow
  |
  = note: `#[deny(arithmetic_overflow)]` on by default
```

The original code comes from macro-generated test cases in `num-integer`, which failed here:
https://github.com/rust-num/num-integer/pull/54#issuecomment-1488962831

### Version it worked on

It most recently worked on: 1.69.0-beta.3

### Version with regression

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (478cbb42b 2023-03-28)
binary: rustc
commit-hash: 478cbb42b730ba4739351b72ce2aa928e78e2f81
commit-date: 2023-03-28
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

**Comment by cuviper** [comment]
Note that this only errors in a full build, not check.

`cargo bisect-rustc` identified b05bb2900825e4d063ebf3f0f11afb08baed3bc6 -- cc #10887 @cjgillot @oli-obk

**Comment by cuviper** [comment]
@cjgillot I found that this is also failing in some num-traits [test cases](https://github.com/rust-num/num-traits/blob/b5906eeeab1fffd1458285c036e099be01d8cd14/tests/cast.rs#L242-L292), and the new change doesn't help. Here's a minimization of that:

```rust
use std::cmp::Ordering::*;
fn main() {
    let max = u64::MAX as usize;
    match (usize::MAX as u64).cmp(&u64::MAX) {
        Greater => {
            dbg!(max + 1);
        }
        Equal | Less => {}
    }
}
```

```
$ rustc +nightly -V
rustc 1.70.0-nightly (28a29282f 2023-04-06)
$ rustc +nightly test.rs
error: this arithmetic operation will overflow
 --> test.rs:6:18
  |
6 |             dbg!(max + 1);
  |                  ^^^^^^^ attempt to compute `usize::MAX + 1_usize`, which would overflow
  |
  = note: `#[deny(arithmetic_overflow)]` on by default
```

This doesn't error if the `Equal | Less` is changed to `_`, or if the `match` is changed to `if let`.

I worry that this is a weakness in determining what is unreachable, especially if we were to expand that scenario to something that is not *statically* unreachable, only realized at runtime.

**Comment by cjgillot** [comment]
@cuviper the existing lint does not try to call u64::cmp, so it conservatively assumes that it can return anything, so that the 3 branches are reachable. 

Removing false positives from this lint is inherently limited by the halting problem, and pragmatically by the resources we accept to put in the symbolic execution.

I don't think keeping this bug open is very useful, except to redirect possible duplicates. I fixed the previous instance because the 'if false' can realistically happen with consts. There are still 1001 variations that still have the false positive. 

Should we suggest to allow the lint?

**Comment by cuviper** [comment]
IMO, all of this symbolic execution should still behave [as-if](https://en.wikipedia.org/wiki/As-if_rule) it were executed at runtime. Introducing a compile-time error is going too far, and worse it doesn't respect `-Coverflow-checks` either, where it might not have even been a runtime error even if it were totally reachable.

But I have no idea how hard it would be to implement as-if behavior here. Has the current behavior already been discussed and decided, or should we nominate this for compiler discussion?

> Should we suggest to allow the lint?

Do you mean suggest to the maintainer of the num crates? That's me, so we don't need to take any further action than this issue. If the change stands, I can and will adapt as necessary.

If you mean in general, having the compiler suggest to allow its own lint, I think that would be a bad outcome. If there are 1001 variations of false positives in a deny-by-default lint, we have a problem.

**Comment by cuviper** [comment]
BTW, I noticed that each time I try one of these failing tests, it's leaving temporary `*.rcgu.o` files around. That may have nothing to do with your changes, but it seems like this lint is not cleaning up properly. Should I file another issue?

**Comment by pnkfelix** [comment]
Discussed [briefly in T-compiler triage meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-13/near/349134541), but we did not really make progress on trying to help towards a decision here...

**Comment by jyn514** [comment]
Yes, I don't think we should *try* to warn on unreachable code, it seems ok to silence the lint there. I just don't think we should treat it as a regression and breaking change.

**Comment by cuviper** [comment]
While it may be decided that this is not considered a regression, for now this has reached beta, so:

@rustbot label -regression-from-stable-to-nightly +regression-from-stable-to-beta

**Comment by apiraino** [comment]
T-compiler visited this in two meetings on Zulip, posting here the notes for reference: [meeting #1](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-20/near/351345957) and [meeting #2](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-27/near/353335702) - we'll need to circle back on the question when a problem with a lint like this is a bug vs when it is just qualityy of live improvement.

**Comment by apiraino** [comment]
Steering meeting scheduled for discussion: https://github.com/rust-lang/compiler-team/issues/627

@rustbot label -I-compiler-nominated

**Comment by pnkfelix** [comment]
We had the rust-lang/compiler-team#627 meeting, but this wasn't completely settled, at least in terms of whether we need to take action for 1.70 or if we're going to let it ship as is. See https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.5D.202023-05-12.20lint.20bugs.20vs.20QOI.20compiler-team.23627/near/357903013

@rustbot label +I-compiler-nominated

**Comment by pnkfelix** [comment]
based on further discussion in [today's triage meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-05-18/near/359419874) and follow-up chat between Felix and Wesley (as well as [follow-up on last week's steering meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.5D.202023-05-12.20lint.20bugs.20vs.20QOI.20compiler-team.23627/near/359293824)), we're going to let this ride the trains, which means we can remove the nomination label.

@rustbot label: -I-compiler-nominated

**Comment by RodBurman** [comment]
I merged the above code samples into:
```
use std::cmp::Ordering::*;
fn main() {
    // 1
    let x = 2u32;
    let y = 3u32;
    // ...
    if y <= x {
        dbg!(x - y);
    }
    println!("{x} {y}");
    // 2
    let max = u64::MAX as usize;
    match (usize::MAX as u64).cmp(&u64::MAX) {
        Greater => {
            dbg!(max + 1);
        }
        Equal | Less => {}
    }
    println!("{max}");
}
```
and ran both "cargo build" and "cargo clip" against this code, this gave no error or warning messages. This is using:
```
 % cargo -vV
cargo 1.96.0 (30a34c682 2026-05-25)
release: 1.96.0
commit-hash: 30a34c6821b57de0aaec83a901aca39f88f6778c
commit-date: 2026-05-25
host: aarch64-apple-darwin
libgit2: 1.9.2 (sys:0.20.4 vendored)
libcurl: 8.7.1 (sys:0.4.87+curl-8.19.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 3.5.4 30 Sep 2025
os: Mac OS 26.5.1 [64-bit]
```
Given this lack of messages, should we close the issue?

**Comment by RalfJung** [comment]
```rust
use std::cmp::Ordering::*;
fn main() {
    let max = u64::MAX as usize;
    match (usize::MAX as u64).cmp(&u64::MAX) {
        Greater => {
            dbg!(max + 1);
        }
        Equal | Less => {}
    }
}
```
still [reproduces](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=7c934becd95b687c7c2ef0256e6be30e) the error for me on nightly & stable

**Comment by RodBurman** [comment]
Yes for me too, but if I add " println!("{max}");" as in my merged code then it builds without issue, I hope that tells those with deeper knowledge something useful. The original example does build without problems. So some problems have been fixed, but not all answers somehow println! is a workaround of sorts.
